### ISSUE TITLE
BUG: if Promise doesn’t complete, remains waiting (instead of timeout)

### ISSUE THREAD

**Issue by akauppi** [akauppi/firebase-jest-testing#31]
Steps:

Change this line to have a doc ID that doesn't get changed:

```
    await expect( waitForIt("projects/1/userInfo/abc_") ).resolves.toContainObject(william);
```

```
$ npm test
```

Expected:

- test should fail
- user should get back to command line prompt

Actual:

- execution does not return to command line

**Comment by akauppi** [comment]
The problem is that there's no timeout in the Promise, and JavaScript promises cannot be cancelled from outside.

**Comment by akauppi** [comment]
Suggestion for a work-around:

If we’re able to detect when the Jest environment has been torn down, make a `setTimeout` callback check that every 100ms or so, and cancel the Promise ”from within” if the environment is down.

This won’t allow us to change a timeout (in the reports) to success/fail, but it should be enough to let the execution pass (focus of the issue).

…aaand we can know the environment is down simply by making an `afterAll`

**Comment by akauppi** [comment]
https://github.com/facebook/jest/issues/9685

**Comment by akauppi** [comment]
Could bring `eventually` back, maybe as:

`eventually.doc(docPath)`

**Comment by akauppi** [comment]
Did a small experiment; can indeed tap into the `afterAll` and have the execution return to OS. (however...)

The thing keeping Jest from returning the control is the Firebase listener; not an unresolved Promise (Jest seems fine with those).

A suitable solution can be done. Since this repo is about Firebase / Jest interaction, this might be as good a place as any to do it.

(ultimately, Firebase could be more Jest friendly?)

**Comment by akauppi** [comment]
Fixed in 05b1eb9e4e84dc9189e40021814c59a8ee19cd78, coming to next release.

The fix wraps `.onSnapshot` so that the programming API remains the normal for tests, but if there are unreleased listeners when Jest exits, those are released.
